apisix/core/config_etcd.lua

```diff
@@ -188,9 +190,43 @@ local function run_watch(premature)
 
         ::watch_event::
         while true do
-            local res, err = res_func()
-            if log_level >= NGX_INFO then
-                log.info("res_func: ", inspect(res))
+            local res
+            local sema, err = semaphore.new()
+            if not sema then
+                log.error("create sema failed", err)
+                goto watch_event
+            end
+
+            local get_res_th = ngx_thread_spawn(function ()
+                res, err = res_func()
+                if log_level >= NGX_INFO then
+                    log.info("res_func: ", inspect(res))
+                end
+                sema:post()
+            end)
+
+            local check_worker_th = ngx_thread_spawn(function ()
+                while not exiting() do
+                    ngx_sleep(0.1)
+                end
+                sema:post()
+            end)
+
+            local _, sem_err = sema:wait(opts.timeout + 10)
```

Comment:
Could we use ngx.thread.wait instead of sema?

Reply:
Great suggestion, I didn't notice this method before, I will update it.